### PR TITLE
Using dynamic url for dashboard state transitions and adding it to web-cloud

### PR DESCRIPTION
**Pull request by AdityaHegde**

### CODE REVIEW

File: web-common/src/features/dashboards/proto-state/StateSyncManager.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  public handleUrlChange() {
+    const pageUrl = get(page).url;
+    const newUrlState = pageUrl.searchParams.get("state");
+    if (this.urlState === newUrlState) return;
+    this.urlState = newUrlState;
+
+    // run sync if we didn't change the url through a state change
+    // this can happen when url is updated directly by the user
+    if (!this.updating && this.urlState && this.urlState !== this.protoState) {
+      const partialDashboard = fromUrl(pageUrl);
```
Comment (ericpgreen2) [inline web-common/src/features/dashboards/proto-state/StateSyncManager.ts:36]: I find both the `partialDashboard` and the `fromUrl` names a bit hard to follow

**Review by ericpgreen2 — APPROVED** [review]
Left a few suggestions, but looks to me like it does the job. I'll approve, though just realizing you tagged Benjamin for a review.

File: web-common/src/features/dashboards/proto-state/StateSyncManager.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    this.protoState = metricsExplorer.proto;
+
+    // if state didn't change do not call goto. this avoids adding unnecessary urls to history stack
+    if (this.protoState !== this.urlState) {
+      goto(`${pageUrl.pathname}?state=${this.protoState}`);
+      this.updating = true;
+    }
+  }
+
+  public handleUrlChange() {
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-common/src/features/dashboards/proto-state/StateSyncManager.ts:27]: Its more of a convenience. Where ever this is used we would need another wrapper function to make sure the reactive statement doesnt run unnecessarily. That is the reason this doesnt take the url. Unlike the other method where it doesnt matter.

File: web-admin/src/routes/-/[organization]/[project]/dashboard/[name]/+page.svelte [hunk 3]
```diff
@@ -1,9 +1,20 @@
...
   import { page } from "$app/stores";
   import { Dashboard } from "@rilldata/web-common/features/dashboards";
+  import { useDashboardStore } from "@rilldata/web-common/features/dashboards/dashboard-stores";
+  import { StateSyncManager } from "@rilldata/web-common/features/dashboards/proto-state/StateSyncManager";
 
   const metricViewName: string = $page.params.name;
-  $: console.log("metricViewName", metricViewName);
+  const stateSyncManager = new StateSyncManager(metricViewName);
+
+  $: metricsExplorer = useDashboardStore(metricViewName);
```
Comment (ericpgreen2) [inline web-admin/src/routes/-/[organization]/[project]/dashboard/[name]/+page.svelte:10]: Now for consistency, I think it makes sense to switch this from `metricsExplorer` to `dashboardStore`

File: web-common/src/features/dashboards/proto-state/StateSyncManager.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ericpgreen2) [reply web-common/src/features/dashboards/proto-state/StateSyncManager.ts:27]: Ah I just figured it'd be cleaner to get closer to a pure function

File: web-admin/src/routes/-/[organization]/[project]/dashboard/[name]/+page.svelte (hunk 3 above)
Comment (AdityaHegde) [reply web-admin/src/routes/-/[organization]/[project]/dashboard/[name]/+page.svelte:10]: honestly there is too much inconsistancy in the naming in dashboard feature. It is called `metricsExplorer` in all components. So lets refactor that in one shot.

File: web-common/src/features/dashboards/proto-state/StateSyncManager.ts [hunk 4]
```diff
@@ -33,9 +33,12 @@ export class StateSyncManager {
     // run sync if we didn't change the url through a state change
     // this can happen when url is updated directly by the user
     if (!this.updating && this.urlState && this.urlState !== this.protoState) {
-      const partialDashboard = fromUrl(pageUrl);
-      if (partialDashboard) {
-        metricsExplorerStore.syncFromUrl(this.metricViewName, partialDashboard);
+      const partialDashboardState = getDashboardStateFromUrl(pageUrl);
```
[the remark this answers is not in the document]
Comment (AdityaHegde) [reply web-common/src/features/dashboards/proto-state/StateSyncManager.ts:38]: added comments

File: web-admin/src/routes/-/[organization]/[project]/dashboard/[name]/+page.svelte (hunk 3 above)
Comment (ericpgreen2) [reply web-admin/src/routes/-/[organization]/[project]/dashboard/[name]/+page.svelte:10]: Agreed we're inconsistent right now. I was thinking new code could use the "dashboard" terminology and over time we'd transition to the right state.

**Comment by ericpgreen2** [comment]
I just played around with this branch in a `web-admin` project, I interacted with a dashboard, and the URL didn't update. It never added a `state` query parameter. @AdityaHegde, can you test it on `web-admin` too?

**Review by ericpgreen2 — APPROVED** [review]
Per Slack discussion w/ Hamiltion, approving so the `web-admin` issues don't block this from fixing our `web-local` problems. `web-admin` can be addressed in a follow-on.
